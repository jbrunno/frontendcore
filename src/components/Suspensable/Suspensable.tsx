import React, { ComponentType, lazy, Suspense } from 'react';
import { ErrorBoundary, FallbackProps } from 'react-error-boundary';

function ErrorFallback({ error }: FallbackProps) {
  return (
    <div>
      <p>Something went wrong:</p>
      <p>{error.message}</p>
    </div>
  );
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type SuspensableComponent<C = ComponentType<{}>> =
  | C
  | Promise<{ default: C }>;

type SuspensableProps = {
  component: SuspensableComponent;
  suspense?: boolean;
  children: JSX.Element;
};

export function Suspensable({
  children,
  suspense = true,
  component,
}: SuspensableProps) {
  const Component =
    component instanceof Promise ? lazy(() => component) : component;

  if (suspense) {
    return (
      <Suspense fallback={<div>Loading...</div>}>
        <ErrorBoundary FallbackComponent={ErrorFallback}>
          <Component />
          {children}
        </ErrorBoundary>
      </Suspense>
    );
  }

  return (
    <>
      <Component />
      {children}
    </>
  );
}
