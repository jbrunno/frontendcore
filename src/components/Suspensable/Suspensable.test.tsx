import React, { ComponentType } from 'react';
import { render } from '@test-utils/component-testing';
import { Suspensable } from './Suspensable';

test('renders the Suspensable', () => {
  const promise = new Promise<{ default: ComponentType }>((resolve) => {
    setTimeout(() => {
      resolve({ default: () => <div>Test</div> });
    }, 1000);
  });
  const { container } = render(
    <Suspensable component={React.lazy(() => promise)}>
      <div>children</div>
    </Suspensable>,
  );

  expect(container).toMatchSnapshot();
});
