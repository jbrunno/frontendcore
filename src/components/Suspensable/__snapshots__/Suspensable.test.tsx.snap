// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Suspensable 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
